x-pack/plugins/console_extensions/server/plugin.ts

```diff
@@ -3,30 +3,90 @@
  * or more contributor license agreements. Licensed under the Elastic License;
  * you may not use this file except in compliance with the Elastic License.
  */
-import { join } from 'path';
-import { CoreSetup, Logger, Plugin, PluginInitializerContext } from 'kibana/server';
+import {
+  CoreSetup,
+  CoreStart,
+  Logger,
+  Plugin,
+  PluginInitializerContext,
+  SavedObjectsClientContract,
+  SavedObjectsClient,
+} from '../../../../src/core/server';
+
+import { APP } from '../common/constants';
 
 import { ConsoleSetup } from '../../../../src/plugins/console/server';
+import { PluginSetupContract as FeatureSetup } from '../../features/server';
+import { SecurityPluginSetup } from '../../security/server';
+
+import { installExtensions } from './spec/install_extensions';
 
-import { processors } from './spec/ingest/index';
+import { registerTextObjectsRoutes } from './routes/api/text_object';
 
 interface SetupDependencies {
   console: ConsoleSetup;
+  features: FeatureSetup;
+  security: SecurityPluginSetup;
 }
 
 export class ConsoleExtensionsServerPlugin implements Plugin<void, void, SetupDependencies> {
-  log: Logger;
+  private internalSavedObjectsClient: SavedObjectsClientContract | null = null;
+
+  private readonly log: Logger;
+
   constructor(private readonly ctx: PluginInitializerContext) {
     this.log = this.ctx.logger.get();
   }
 
   setup(
-    core: CoreSetup,
-    { console: { addProcessorDefinition, addExtensionSpecFilePath } }: SetupDependencies
+    { http, getStartServices }: CoreSetup,
+    { console, features, security: { authc } }: SetupDependencies
   ) {
-    addExtensionSpecFilePath(join(__dirname, 'spec/'));
-    processors.forEach(processor => addProcessorDefinition(processor));
-    this.log.debug('Installed console autocomplete extensions.');
+    // We register this feature explicitly without direct access to Saved Objects
+    // per docs/developer/plugin/development-plugin-feature-registration.asciidoc
+    features.registerFeature({
+      id: APP.id,
+      name: APP.name,
+      app: [],
+      privileges: {
+        all: {
+          savedObject: {
+            all: [],
+            read: [],
+          },
+          ui: [],
+        },
+        read: {
+          savedObject: {
+            all: [],
+            read: [],
+          },
+          ui: [],
+        },
+      },
+    });
+
+    getStartServices().then(([{ savedObjects }]) => {
+      this.internalSavedObjectsClient = new SavedObjectsClient(
+        savedObjects.createInternalRepository(['textObject'])
+      );
+    });
+
+    const router = http.createRouter();
+
+    registerTextObjectsRoutes({
```

Comment:
These routes should have `access:` tags on them so only users who are authorized for Dev Tools have access to them per https://www.elastic.co/guide/en/kibana/current/development-plugin-feature-registration.html

> Unlike the Canvas example above, Dev Tools does not require access to any saved objects to function. Dev Tools does specify an API endpoint, however. When this is configured, the Security plugin will automatically authorize access to any server API route that is tagged with access:console, similar to the following:
> ```
>server.route({
> path: '/api/console/proxy',
> method: 'POST',
> config: {
>   tags: ['access:console'],
>   handler: async (req, h) => {
>     // ...
>   }
> }
>});
> ```

When users have "Read" access to Dev Tools, we might want to restrict them from being able to create the text-objects, as users with read-only access have traditionally been unable to persist any type of saved-object.

Reply:
Makes sense to add the tag :+1:. I am just not certain what a case would be for disallowing certain users from creating saved objects here. We would also not want to let Console save their text in Saved Objects through the UI even with security available?